File: frontend/src/api/report.ts
import apiClient from './client'

// ─── Types ────────────────────────────────────────────────────────────────────

export interface DosenSksRekap {
  dosen_id: string
  dosen_nama: string
  dosen_kode: string
  total_sks: number
  breakdown: Record<string, number>
  bkd_limit_sks: number | null
  bkd_flag: 'ok' | 'near_limit' | 'over_limit' | 'no_limit'
}

export interface SksRekapResponse {
  sesi_id: string
  items: DosenSksRekap[]
  total_dosen: number
}

// ─── API Functions ────────────────────────────────────────────────────────────

export async function getSksRekap(sesiId: string): Promise<SksRekapResponse> {
  const res = await apiClient.get<SksRekapResponse>(`/sesi/${sesiId}/reports/sks-rekap`)
  return res.data
}

// ─── Preferences Summary ──────────────────────────────────────────────────────

export interface DosenPreferenceSummaryItem {
  dosen_id: string
  kode: string
  nama: string
  total_preferensi: number
  total_dilanggar: number
}

export interface PreferencesSummaryResponse {
  sesi_id: string
  total_preferensi: number
  total_dilanggar: number
  breakdown: DosenPreferenceSummaryItem[]
}

export async function getPreferencesSummary(sesiId: string): Promise<PreferencesSummaryResponse> {
  const res = await apiClient.get<PreferencesSummaryResponse>(`/sesi/${sesiId}/preferences-summary`)
  return res.data
}

// ─── Room Map ─────────────────────────────────────────────────────────────────

export interface RoomCellInfo {
  kode_mk: string
  nama_mk: string
  kelas: string | null
  dosen: string
}

export interface RoomMapSlot {
  hari: string
  sesi: number
  label: string
  rooms: Record<string, RoomCellInfo | null>
}

export interface RoomMapResponse {
  sesi_id: string
  rooms: string[]
  days: string[]
  slots: RoomMapSlot[]
}

export async function getRoomMap(sesiId: string): Promise<RoomMapResponse> {
  const res = await apiClient.get<RoomMapResponse>(`/sesi/${sesiId}/reports/room-map`)
  return res.data
}

File: backend/app/schemas/report.py
"""
backend/app/schemas/report.py
Pydantic v2 schemas untuk endpoint laporan (reports).
"""

import uuid
from typing import Optional

from pydantic import BaseModel


class DosenSksRekap(BaseModel):
    dosen_id: uuid.UUID
    dosen_nama: str
    dosen_kode: str
    total_sks: int
    breakdown: dict[str, int]
    bkd_limit_sks: Optional[int]
    bkd_flag: str  # "ok" | "near_limit" | "over_limit" | "no_limit"


class SksRekapResponse(BaseModel):
    sesi_id: uuid.UUID
    items: list[DosenSksRekap]
    total_dosen: int


# ---------------------------------------------------------------------------
# Room-map schemas
# ---------------------------------------------------------------------------

class RoomCellInfo(BaseModel):
    """Informasi MK yang menempati sel ruang × timeslot."""
    kode_mk: str
    nama_mk: str
    kelas: Optional[str]
    dosen: str  # nama dosen1 (+ dosen2 jika ada)


class RoomMapSlot(BaseModel):
    """Satu baris dalam matrix: satu timeslot dengan semua ruang."""
    hari: str
    sesi: int
    label: str
    rooms: dict[str, Optional[RoomCellInfo]]  # nama_ruang → info atau None


class RoomMapResponse(BaseModel):
    sesi_id: uuid.UUID
    rooms: list[str]          # daftar nama ruang aktif (urut)
    days: list[str]           # daftar hari unik (urut)
    slots: list[RoomMapSlot]  # semua 15 timeslot dengan isi per ruang
